### PR TITLE
Fix/WC - handling lost connections in wallet connect

### PR DESCRIPTION
**Pull request by alexjba**
### What does the PR do

Closes: #16180 #15636

1. Implementing the session events whenever Status loses connection to chains
2. Handling the integration when there is no internet connection

Not included:
2. Polished UI - needs design

<details><summary>WC/EIP documentation for the `connect` and `disconnect` events.
</summary>
<p>
Note: This is also covered by EIP documentation https://eips.ethereum.org/EIPS/eip-1193

This event can signal the following events:

connect
The Provider emits connect when it:

first connects to a chain after being initialized.
first connects to a chain, after the disconnect event was emitted.
interface ProviderConnectInfo {
readonly chainId: string;
}
The event emits an object with a hexadecimal string chainId per the eth_chainId Ethereum RPC method, and other properties as determined by the Provider.

disconnect
The Provider emits disconnect when it becomes disconnected from all chains.
This event emits a [ProviderRpcError](https://eips.ethereum.org/EIPS/eip-1193#errors). The error code follows the table of [CloseEvent status codes](https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent#Status_codes).

</p>
</details> 

In the new flow there are 2 main components involved

- `ui/app/AppLayouts/Wallet/helpers/ChainsAvailibilityWatchdog.qml`. Component watching the chains availability and network availability
- `ui/app/AppLayouts/Wallet/services/dapps/plugins/ChainsSupervisorPlugin.qml`. Component reacting on chains availability changes by sending the appropriate events to the sdk.

Commits:

- https://github.com/status-im/status-desktop/commit/000d3c5d6cfddc6ed60521ad0358fe9fd5aa5e85 - A small fix in nim to optimise the qml signals whenever a chain online state changes. Here we've had a QObject exposed from nim with properties like `blockedNetworks`. The `changed` event on the exposed `QObject` was being sent when the underlying data might or might not have been changed. This was not needed as the underlying property emits proper signals.

- https://github.com/status-im/status-desktop/commit/3028034759c3b1b838803c66b5f21224823eeca1 - Exposing `emitSessionEvent` to qml

- https://github.com/status-im/status-desktop/commit/b552f2483e64fcfc87fbb9ca2f2b3748a073ffc9 - Add an option to fully disable any button that triggers a WC interaction from the UI

- https://github.com/status-im/status-desktop/commit/16fc962c8e7d65ed170f537922f23bc86ae184fd - Adding the `ChainsAvailibilityWatchdog.qml ` and `ChainsSupervisorPlugin.qml ` components to handle the network or chain connection loss. These events will send the `connect`, `disconnect` session events to WC when needed.

### Affected areas

WalletConnect

### Architecture compliance

- [x] I am familiar with the application architecture and agreed good practices.
My PR is consistent with this document: [Status Desktop Architecture Guide](https://github.com/status-im/status-desktop/blob/master/CONTRIBUTING.md)

### Screenshot of functionality (including design for comparison)

https://github.com/user-attachments/assets/fbf4675f-1135-4ef8-9d66-551c30f025c8

https://github.com/user-attachments/assets/25fb48e4-aca1-4593-a08f-112dc1645441

- [x] I've checked the design and this PR matches it

### CODE REVIEW

File: ui/app/AppLayouts/Wallet/services/dapps/plugins/ChainsSupervisorPlugin.qml [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                        session.namespaces.eip155.chains.forEach((chainId) => {
+                            const chainInt = DAppsHelpers.chainIdFromEip155(chainId)
+                            if (!d.sessionHasChainSupport(session, chainInt)) {
+                                return
+                            }
+
+                            sdk.emitSessionEvent(session.topic, {
+                                name: "disconnect",
+                                data: {
+                                    code: 4900,
```
Comment (caybro) [inline ui/app/AppLayouts/Wallet/services/dapps/plugins/ChainsSupervisorPlugin.qml:113]: Magic constant? :)

**Review by caybro — APPROVED** [review]
LGTM in general

File: ui/app/AppLayouts/Wallet/services/dapps/WalletConnectSDK.qml [hunk 2]
```diff
@@ -244,6 +248,18 @@ WalletConnectSDKBase {
             )
         }
 
+        function emitSessionEvent(topic, event, chainId) {
+            console.debug(`WC WalletConnectSDK.wcCall.emitSessionEvent; topic: ${topic}, event: ${JSON.stringify(event)}, chainId: ${chainId}`)
```
[the remark this answers is not in the document]
Comment (alexjba) [reply ui/app/AppLayouts/Wallet/services/dapps/WalletConnectSDK.qml:252]: yes. We log every entry point in the WC js sdk.

**Comment by alexjba** [comment]
Added qml tests as well

**Comment by alexjba** [comment]
> Added qml tests as well

These QML tests are totally broken on linux 5.15.2. Or maybe it's just the CI configuration as I see some SSL warnings on the failing tests

**Comment by alexjba** [comment]
Works on Mac with 5.15.2. Must be something else

**Comment by caybro** [comment]
> Works on Mac with 5.15.2. Must be something else

The ui-tests failures seem legit:
image

**Comment by alexjba** [comment]
> > Works on Mac with 5.15.2. Must be something else
> 
> The ui-tests failures seem legit: image

Yes, it's a linux 5.15.2 thing..Looking into it

**Comment by alexjba** [comment]
It can't find the openSSL dependency

File: ui/StatusQ/src/networkchecker.cpp [hunk 3]
```diff
@@ -16,7 +16,7 @@ bool NetworkChecker::isOnline() const
 
 void NetworkChecker::checkNetwork()
 {
-    QNetworkRequest request(QUrl("https://fedoraproject.org/static/hotspot.txt"));
+    QNetworkRequest request(QUrl("http://fedoraproject.org/static/hotspot.txt"));
```
Comment (caybro) [inline ui/StatusQ/src/networkchecker.cpp:19]:
```suggestion
    QNetworkRequest request(QUrl(QStringLiteral("http://fedoraproject.org/static/hotspot.txt")));
```

While you're at it :) Btw why the change? Is https not available for you or just slower?

Comment (caybro) [reply ui/StatusQ/src/networkchecker.cpp:19]: Ah, trying to work around the test failure? :thinking:

Comment (alexjba) [reply ui/StatusQ/src/networkchecker.cpp:19]: yes. And it doesn't make sense to me to use the https for this purpose.

**Review by saledjenic — APPROVED** [review]
